Composable scalar transforms

Ickaser opened the pull request:
This is very much still WIP, but took a stab at constructing something like discussed in #95 and #126 . As of yet, I haven't even tested that this runs, much less correctly; will come back when I have more time.

One major question: for things like mapping to negative reals, I am imagining `Scale(-1) ∘ Exp()`. In this case, should the log-Jacobian of `Scale` be `log(abs(scale))`?

Ickaser commented:
I've added some scalar consistency tests now, which pass, with this new API. If there are some other obvious tests, I am open to suggestions. Still missing is any treatment of non-`Real` number types, as would be necessary for Unitful, but if you want to merge this and get feedback first before tackling the Unitful-friendly part this might be ready for something like that.

New questions
- Namespace: to avoid name collisions, I decided to rename `Exp`, `Scale`, and `Shift` to `TVExp`, `TVScale`, and `TVShift`. That way they can be exported and maintain a shorter name than `TransformVariables.Exp`. I'm of course open to other prefixes or ways to keep it readable.
- Should unit transforming use the `Scale` operator (so allow Scale to take non-`Real` types), or give it its own separate transformation (which I guess could accommodate other kinds of exotic number types)?

Ickaser commented:
Some other ideas I had but which haven't tried out yet (note to self as much as anything else)
- We could try replacing the existing `ShiftedExp` and `ScaledShiftedLogistic` with these elemental transforms, but there may be some error/sanity checks we miss out on because a `CompositeScalarTransform` doesn't yet have end-to-end error checking like these manually-written ones do.
- I haven't yet tried composing prior transformations with the new ones. If I'm lucky it will just work
- It may make sense to introduce a `TVNeg()` transformation for a simple sign change, rather than using e.g. `TVScale(-1)` which encodes a number type in the `-1`.

Ickaser commented:
I decided to add a `TVNeg` transformation, which if I am understanding correctly has a log-Jacobian of 0, and enforce that `TVScale` takes only positive numbers. Some tests are also in place and pass on my machine. I haven't gone so far as to replace existing transforms with the refactored form, which would put the new version through the full test suite.

I haven't thought deeply about what this current implementation will do with real number types, e.g. `Float32`, either in parameters or inputs.

As far as using this to introduce Unitful support, I see two main options:
1. Create a Unitful extension with a single scalar transform, e.g. `TVUnit(u::Unitful.Units)`. This transform would construct a `Quantity` in the forward direction, and apply `ustrip` for the inverse transform. This approach would have to be custom-implemented for each exotic number types (e.g. units provided by `DynamicQuantities`), but I think there would be a natural way to do so for any given type.
2. Generalize the `TVScale` transform to allow any object; in this case, you multiply by units in the forward direction, then the inverse divides by the units. In the case of Unitful, this should produce the right types on the inverse. For other number types, multiplying should yield new types with the help of promotion, but the inverse would not be so simple.

In case 1, new number types each need their own extension. In case 2, inverse transforms will not yield original types.

tpapp commented:
I think this is shaping up nicely, thanks for the work! Sorry I have been very busy with a project and failed to provide any feedback so far, I will try to make up for it now.

>  I am imagining Scale(-1) ∘ Exp(). In this case, should the log-Jacobian of Scale be log(abs(scale))?

Chaining just multiplies the Jacobians, so abs(log(...)) Jacobians are just added up. It should be something like `log(abs(scale)) + x`, I think your implementation is correct.

> In case 2, inverse transforms will not yield original types.

I think that is OK, I think we should just clarify the API for `inverse`. In particular, the best the user can expect is a vector of equal length, approximately equal elementwise. When this PR is done I will make that clear.

Ickaser commented:
I went ahead and tried out replacing the `ShiftedExp` and `ScaledShiftedLogistic` with the corresponding refactored transformations, and by and large the refactor works with the existing test suite. Exceptions that I marked with `@test_broken`:
- `as(Real, 1, 4.0) == as(Real, 1.0, 4.0)` 
  - The new transforms each hold their own number types (e.g. `TVShift{Int64}` vs `TVShift{Float64}`.  Since my goal is to eventually accommodate new (unitful) number types, I don't know whether to continue to enforce this. I can define something like `==(TVShift(a), TVShift(b)) = (a==b)` and run with that
- Scalar show tests. 
  - This is a valid question: I decided to show a `CompositeScalarTransform` by printing a valid expression (`TVShift(5) ∘ TVExp()`); would you prefer to represent it differently? Should there be special dispatches for e.g. `asℝ₋`, which is `TVNeg() ∘ TVExp()`, like there were for `ShiftedExp` and `ShiftedScaledLogistic`? In particular, if we have a special show for `asℝ₊ == TVExp()`, then we get results like `TVExp() ∘ TVScale(2)` printed like `asℝ₊ ∘ TVScale(2)`, which I think I don't like.

Ickaser commented:
> > In case 2, inverse transforms will not yield original types.
> 
> I think that is OK, I think we should just clarify the API for `inverse`. In particular, the best the user can expect is a vector of equal length, approximately equal elementwise. When this PR is done I will make that clear.

I guess if we take this approach and have a `TVScale{T}` with `T` that can take any type, it is always possible to make an extension with dispatches like `transform(t::TVScale{T}, x) where T <: DynamicQuantities.Quantity`. So the two options are not mutually exclusive, and option 2 provides a hopefully-sane fallback anyway.

Ickaser commented:
Okay, tried widening types, etc. to accomodate Unitful types. For now, that means that `transform` accepts `Real`s, so non-`Real` numbers have to show up only in the outermost transform; `inverse` accepts all `Number`s, of which `Unitful.Quantity` is a subtype.

A quick check suggests that `transform` and `inverse` work fine, but `transform_and_logjac` does not. Frankly, I have no idea how to treat the log-Jacobian for scaling by a unit--the Jacobian in that case should have units, I think, which would mean the log-Jacobian would require taking the log of a dimensioned quantity (which yields a Julia error, but is also not well-defined).

Should that be explicitly documented as not working for dimensional quantities? Is there a different way to treat the log-Jacobian?

Ickaser commented:
Okay, I've implemented it that way and added a note to the docs.

I have `inverse_and_logjac` defined for all of the individual scalar transforms, as well as the composite; should I move that from its own WIP place in the test suite to get tested with the rest of the scalar consistency tests?

With that, I'm not thinking of anything else that's missing from this implementation. Is there anything still on your list? Does it need more documentation or anything? Should we delete the `ShiftedExp` and `ScaledShiftedLogistic`, now that they have been superceded by this implementation?

Ickaser commented:
I went ahead and added an alias `TransformVariables.compose` for `Base.:∘`, because Unicode is still sometimes annoying to type. Let me know if you don't like that option.

I've reworked the scalar show behavior so that all the old show tests still work. With that, I don't think we're breaking any API, just adding new features.

I think everything on my list is working at this point. Anything else you want to see in this?

Ickaser commented:
(Remaining question: should we delete the `ShiftedExp` and `ScaledShiftedLogistic`, since they are not used in any public API nor in any of the internal constructions?)

tpapp commented on docs/src/index.md:
````diff
@@ -109,11 +109,12 @@ TVNeg
 ```
 
 Consistent with common notation, transforms are applied right-to-left; for example, `as(Real, ∞, 3)` is equivalent to `TVShift(3) ∘ TVNeg() ∘ TVExp()`.
+If you are working in an editor where typing Unitful is difficult, `TransformVariables.compose` (not exported) is provided as an alias for `∘`, as in `TransformVariables.compose(TVScale(5.0), TVExp(), TVNeg())`.
````
perhaos "typing Unitful" was meant to be "typing Unicode"?

Ickaser replied:
yep

tpapp commented on src/scalar.jl:
```diff
@@ -181,6 +181,7 @@ Base.:∘(t::ScalarTransform, ct::CompositeScalarTransform) = CompositeScalarTra
 Base.:∘(ct::CompositeScalarTransform, t::ScalarTransform) = CompositeScalarTransform((ct.transforms..., t))
 Base.:∘(ct1::CompositeScalarTransform, ct2::CompositeScalarTransform) = CompositeScalarTransform((ct1.transforms..., ct2.transforms...))
 Base.:∘(t::ScalarTransform, tt::Vararg{ScalarTransform}) = CompositeScalarTransform((t, tt...))
+const compose = Base.:∘
```
I wonder if aliasing the whole of `∘` makes sense here; perhaps just alias and `CompositeScalarTransform`?

Ickaser replied:
Good call, it was trivial to replace this with a direct constructor.

tpapp commented:
> remaining question: should we delete the ShiftedExp and ScaledShiftedLogistic, since they are not used in any public API nor in any of the internal constructions?

1. we should benchmark first, to see if they are equally effective,
2. if yes, deprecate current ones in a minor release,
3. then remove in the next major release.

> I think everything on my list is working at this point. Anything else you want to see in this?

Sorry, I had computer issues and I am still catching up with FOSS work, will try to review the whole thing ASAP.

Ickaser commented:
> > remaining question: should we delete the ShiftedExp and ScaledShiftedLogistic, since they are not used in any public API nor in any of the internal constructions?
>     
> 1. we should benchmark first, to see if they are equally effective,
> 
> 2. if yes, deprecate current ones in a minor release,
> 
> 3. then remove in the next major release.

1. Benchmark in terms of performance (run time), I take it? I will try to put together a little script for that purpose.
To make sure I understand 2 and 3, we will:
2. In the minor release, make available the new composite transform functionality, but have `as(Real, a, b)` construct a `ShiftedExp` or `ScaledShiftedLogistic` as before
3. In the major release, have `as(Real, a, b)` use the composite transforms instead

This PR already makes `as(Real, a, b)` use the composite transforms, so I will have to move some of the current unit tests and changes to a new PR so it can get applied for a major release. But this makes sense to me as a deprecation path, so (if I have understood the idea correctly) I am willing to do that.

tpapp commented on test/runtests.jl:
```diff
@@ -46,20 +47,96 @@ end
+        t1 = TVScale(2u"m") ∘ TVShift(a) ∘ TVExp() 
+        test_transformation(t1, y -> y > a*2u"m", jac=false)
+        t2 = TVScale(1u"s") ∘ TVShift(a) ∘ TVScale(b-a) ∘ TVLogistic()
+        test_transformation(t2, y -> (a*u"s" < y < b*u"s"), jac=false)
```
maybe a test that it errors for `transform_and_logjac`

Ickaser replied:
I did have this in another test set, but if you naturally expected that here, this is probably the more logical place. Moved to here

tpapp commented on test/runtests.jl:
```diff
@@ -46,20 +47,96 @@ end
 @testset "scalar transformation corner cases" begin
     @test_throws ArgumentError as(Real, "a fish", 9)
-    @test as(Real, 1, 4.0) == as(Real, 1.0, 4.0)
+    @test_broken as(Real, 1, 4.0) == as(Real, 1.0, 4.0)
```
you might as well remove this test

tpapp commented on test/utilities.jl:
```diff
@@ -31,19 +31,21 @@ automatic differentiation.
 """
 function test_transformation(t::AbstractTransform, is_valid_y;
-                             vec_y = identity, N = 1000, test_inverse = true)
+                             vec_y = identity, N = 1000, test_inverse = true, jac=true)
```
please document `jac` in the docstring above

tpapp reviewed (changes requested):
Thanks for all the fixes, this is shaping up nicely so just minor comments. If benchmarks show no regression, this is good to go.

tpapp commented:
> make available the new composite transform functionality, but have as(Real, a, b) construct a ShiftedExp or ScaledShiftedLogistic as before

Sorry, I take it back, I forgot that we never exported `ShiftedExp` etc, so no need to deprecate.

If the benchmarks show no regression, I think this is good to go.

Ickaser commented:
Pretty simple benchmark, let me know if this is incomplete (first time really investigating something like this):

```julia
using TransformVariables
using BenchmarkTools
const TV = TransformVariables
btrans(t) = inverse(t, transform(t, rand()))
blogjac(t) = transform_and_logjac(t, rand())
```

```julia
t1 = TV.ShiftedExp(true, 4.0)
t2 = as(Real, 4.0, ∞)
@benchmark btrans(t1)
@benchmark btrans(t2)
```
```
BenchmarkTools.Trial: 10000 samples with 991 evaluations per sample.
 Range (min … max):  38.143 ns …  7.402 μs  ┊ GC (min … max): 0.00% … 98.79%
 Time  (median):     40.666 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   47.260 ns ± 83.987 ns  ┊ GC (mean ± σ):  2.86% ±  1.94%

  █▄▆▅▄▃▂▂▁  ▁▁                                               ▁
  █████████▇▇█████▇▆▅▇▇▆▇▇▇▇▇█▆▇▆▇▇▇▆▆▇▇▇▇▆▆▆▆▄▄▆▅▅▅▄▆▅▅▅▅▂▄▄ █
  38.1 ns      Histogram: log(frequency) by time       123 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.

BenchmarkTools.Trial: 10000 samples with 993 evaluations per sample.
 Range (min … max):  36.556 ns …  7.574 μs  ┊ GC (min … max): 0.00% … 99.01%
 Time  (median):     40.282 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   47.887 ns ± 84.888 ns  ┊ GC (mean ± σ):  2.73% ±  1.94%

  █▄▇▄▃▃▂▁▁▂▁▁                                                ▁
  ███████████████▇▇▇▇█▇▇▇██▇▇██▇▇▇▇▆▆▆▅▆▆▆▅▆▆▆▆▇▅▄▄▅▄▆▅▄▅▅▃▄▄ █
  36.6 ns      Histogram: log(frequency) by time       145 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.
```

The other benchmarks 
```julia
@benchmark blogjac(t1)
@benchmark blogjac(t2)

t3 = TV.ShiftedExp(false, 4.0)
t4 = as(Real, -∞, 4.0)
@benchmark btrans(t3)
@benchmark btrans(t4)

t5 = TV.ScaledShiftedLogistic(2.0, 2.0)
t6 = as(Real, 2.0, 4.0)
@benchmark btrans(t5)
@benchmark btrans(t6)
```
all look basically the same, the finite interval takes slightly longer than an exp (as expected) but the two implementations are indistinguishable. 

While I'm thinking about performance, it bears mentioning that, in the test suite, the tests where I compose lots of arbitrary transforms take a lot of time to compile. For example,
```julia
@time begin
    all_transforms = [TVShift(3.0), TVScale(2.0), TVExp(), TVLogistic(), TVNeg(), 
        TV.ScaledShiftedLogistic(2.0, 1.0), TV.ShiftedExp(true, -5), TV.ShiftedExp(false, 4.5)]
    for t1 in all_transforms, t2 in all_transforms, t3 in all_transforms
        t = t1 ∘ t2 ∘ t3
        t isa TransformVariables.CompositeScalarTransform{Tuple{typeof(t1), typeof(t2), typeof(t3)}}
    end
end
```
takes 5 seconds to compile the first time, then something like a millisecond to actually run. I don't know if that's a quirk in how I constructed the test. The tests still only take a couple minutes to run.

devmotion commented on src/scalar.jl:
```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
+"""
+struct CompositeScalarTransform{Ts <: Tuple} <: ScalarTransform
+    transforms::Ts
+    function CompositeScalarTransform(transforms::Ts) where {Ts <: Tuple}
+        @argcheck length(transforms) > 0 DomainError
+        @argcheck all(t -> t isa ScalarTransform, transforms) DomainError
```
Both these checks could be enforced on the type level by e.g. only allowing
```suggestion
    function CompositeScalarTransform(transforms::Ts) where {Ts <: Tuple{ScalarTransform,Vararg{ScalarTransform}}}
```

Ickaser replied:
I think I tried to achieve this but struggled with the syntax, thanks for the help.

devmotion commented on src/scalar.jl:
```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
+Base.firstindex(t::CompositeScalarTransform) = firstindex(t.transforms)
+Base.lastindex(t::CompositeScalarTransform) = lastindex(t.transforms)
+
+transform(t::CompositeScalarTransform, x) = foldr((t, y) -> transform(t, y), t.transforms, init=x)
```
```suggestion
transform(t::CompositeScalarTransform, x) = foldr(transform, t.transforms, init=x)
```

Ickaser replied:
Nice

devmotion commented on src/scalar.jl:
```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
+
+transform(t::CompositeScalarTransform, x) = foldr((t, y) -> transform(t, y), t.transforms, init=x)
+function transform_and_logjac(ts::CompositeScalarTransform, x) 
+    foldr(ts.transforms, init=(x, zero(x))) do t, (x, logjac)
```
```suggestion
    foldr(ts.transforms, init=(x, logjac_zero(LogJac(), typeof(x)))) do t, (x, logjac)
```

devmotion commented on src/scalar.jl:
```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
+function transform_and_logjac(ts::CompositeScalarTransform, x) 
+    foldr(ts.transforms, init=(x, zero(x))) do t, (x, logjac)
+        nx, nlogjac = transform_and_logjac(t, x)
+        x = nx
+        logjac += nlogjac
+        (x, logjac)
```
IMO a bit confusing to reuse the function argument names here:
```suggestion
        (nx, logjac + nlogjac)
```

devmotion commented on src/scalar.jl:
```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
+
+inverse(ts::CompositeScalarTransform, x) = foldl((y, t) -> inverse(t, y), ts.transforms, init=x)
+function inverse_and_logjac(ts::CompositeScalarTransform, x) 
+    foldl(ts.transforms, init=(x, zero(x))) do (x, logjac), t
```
```suggestion
    foldl(ts.transforms, init=(x, logjac_zero(LogJac(), typeof(x)))) do (x, logjac), t
```

devmotion commented on src/scalar.jl:
```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
+function inverse_and_logjac(ts::CompositeScalarTransform, x) 
+    foldl(ts.transforms, init=(x, zero(x))) do (x, logjac), t
+        nx, nlogjac = inverse_and_logjac(t, x)
+        x = nx
+        logjac += nlogjac
+        (x, logjac)
```
```suggestion
        (nx, logjac + nlogjac)
```

devmotion commented on src/scalar.jl:
```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
+as in `ct[i]` or `ct[end]`.
+"""
+struct CompositeScalarTransform{Ts <: Tuple} <: ScalarTransform
+    transforms::Ts
```
I'm not sure about exact trade-offs - but noteably Julia itself uses a different design for function composition:
```suggestion
    outer_transform::T1
    inner_transform::T2
```
Presumably this would also simplify the definitions of `transform` and `inverse` below.

Ickaser replied:
This is a clear case of my just not knowing about more clever algorithms, with recursion like this. But I do like the idea of the transform being a tuple rather than a nested data structure, since it feels easier to reason about (e.g. how to recursively apply from the bottom of a nested struct)

tpapp replied:
Personally I am OK with a `Tuple`, having a preference for flat structures.

devmotion commented on src/scalar.jl:
```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
+Base.:∘(ct::CompositeScalarTransform, t::ScalarTransform) = CompositeScalarTransform((ct.transforms..., t))
+Base.:∘(ct1::CompositeScalarTransform, ct2::CompositeScalarTransform) = CompositeScalarTransform((ct1.transforms..., ct2.transforms...))
+Base.:∘(t::ScalarTransform, tt::Vararg{ScalarTransform}) = CompositeScalarTransform((t, tt...))
+compose(args...) = CompositeScalarTransform(args)
```
This is already achieved by CompositionsBase (https://github.com/JuliaFunctional/CompositionsBase.jl): It defines `compose` as an alias of `∘`, and hence ensures that packages that extend `∘` also share a common non-unicode alternative. Maybe use it instead of defining a different TransformVariables.compose?

Ickaser replied:
Seems like a good idea to me. My question then would be, should `compose` get exported?

devmotion commented on src/scalar.jl:
```diff
@@ -173,21 +311,21 @@ as(::Type{Real}, left, right) =
 
 as(::Type{Real}, ::Infinity{false}, ::Infinity{true}) = Identity()
 
-as(::Type{Real}, left::Real, ::Infinity{true}) = ShiftedExp(true, left)
```
Is there any usecase remaining for `ShiftedExp` etc.? Shouldn't these be removed in the PR?

tpapp replied:
Removal is fine with me.

devmotion commented on src/scalar.jl:
```diff
@@ -173,21 +311,21 @@ as(::Type{Real}, left, right) =
 function as(::Type{Real}, left::Real, right::Real)
     @argcheck left < right "the interval ($(left), $(right)) is empty"
-    ScaledShiftedLogistic(right - left, left)
+    TVShift(left) ∘ TVScale(right - left) ∘ TVLogistic()
```
IMO it would be good to promote `left` and `right` to a common type here - both to reduce the amount of different types (and hence compilation etc.) and to avoid inaccuracies due to less accurate calculations in one part of the chain (e.g., one could think of something extreme like `left::BigFloat` + `right::Int` or less extreme `left::Float32` and `right::Float64`.
```suggestion
    shift, scale = promote(left, right - left)
    TVShift(shift) ∘ TVScale(scale) ∘ TVLogistic()
```

devmotion commented on src/scalar.jl:
```diff
@@ -173,21 +311,21 @@ as(::Type{Real}, left, right) =
 `asℝ₊` and `as_positive_real` are equivalent alternatives.
 """
-const asℝ₊ = as(Real, 0, ∞)
+const asℝ₊ = ∘(TVExp())
```
This seems odd? Why should it be a composite transform?
```suggestion
const asℝ₊ = TVExp()
```

tpapp replied:
I agree, also for the points below.

devmotion commented on src/scalar.jl:
```diff
@@ -205,7 +343,7 @@ Transform to the unit interval `(0, 1)`. See [`as`](@ref).
 `as𝕀` and `as_unit_interval` are equivalent alternatives.
 """
-const as𝕀 = as(Real, 0, 1)
+const as𝕀 = ∘(TVLogistic())
```
```suggestion
const as𝕀 = TVLogistic()
```

tpapp replied:
ditto

devmotion commented on src/scalar.jl:
```diff
@@ -218,6 +356,52 @@ const asℝ = as(Real, -∞, ∞)
 const as_real = asℝ
 
+# Fallback method: print all transforms in order
+function Base.show(io::IO, ct::CompositeScalarTransform)
+    str = string(ct.transforms[1])
+    for ti in ct.transforms[begin+1:end]
+        str *= " ∘ "*string(ti)
+    end
+    print(io, str)
+end
```
It's inefficient to create a string only to print it afterwards. But here it's even simpler, this is just
```suggestion
Base.show(io::IO, ct::CompositeScalarTransform) = join(io, ct.transforms, " ∘ ")
```

Ickaser replied:
Good point!

devmotion commented on src/scalar.jl:
```diff
@@ -218,6 +356,52 @@ const asℝ = as(Real, -∞, ∞)
+end
+
+# If equivalent to asℝ₊, print as such. Two ways to achieve this
+function Base.show(io::IO, ct::CompositeScalarTransform{Tuple{TVShift{T}, TVExp}}) where T
```
```suggestion
function Base.show(io::IO, ct::CompositeScalarTransform{Tuple{<:TVShift, TVExp}})
```

devmotion commented on src/scalar.jl:
```diff
@@ -218,6 +356,52 @@ const asℝ = as(Real, -∞, ∞)
+
+# If equivalent to asℝ₊, print as such. Two ways to achieve this
+function Base.show(io::IO, ct::CompositeScalarTransform{Tuple{TVShift{T}, TVExp}}) where T
+    if ct[1].shift == 0
+        print(io, "asℝ₊")
+    else
+        print(io, "as(Real, ", ct[1].shift, ", ∞)")
+    end
```
IMO that's misleading since with this PR you'd get a transform with a different type with `asℝ₊` - this composite transform might behave similar to `asℝ₊` but it is **not** `asℝ₊`.
```suggestion
    print(io, "as(Real, ", ct.transforms[1].shift, ", ∞)")
```

devmotion commented on src/scalar.jl:
```diff
@@ -218,6 +356,52 @@ const asℝ = as(Real, -∞, ∞)
+end
+Base.show(io::IO, ct::CompositeScalarTransform{Tuple{TVExp}}) = print(io, "asℝ₊")
+
+# If equivalent to asℝ₋, print as such. Two ways to achieve this
+function Base.show(io::IO, ct::CompositeScalarTransform{Tuple{TVShift{T}, TVNeg, TVExp}}) where T
+    if ct[1].shift == 0
+        print(io, "asℝ₋")
+    else
+        print(io, "as(Real, -∞, ", ct[1].shift, ")")
+    end
+end
```
Same here, this seems misleading.

Ickaser replied:
Fair point

devmotion commented on src/scalar.jl:
```diff
@@ -218,6 +356,52 @@ const asℝ = as(Real, -∞, ∞)
+end
+Base.show(io::IO, ct::CompositeScalarTransform{Tuple{TVNeg, TVExp}}) = print(io, "asℝ₋")
+
+# If equivalent to as𝕀, print as such. Two ways to achieve this
+function Base.show(io::IO, ct::CompositeScalarTransform{Tuple{TVShift{T1}, TVScale{T2}, TVLogistic}}) where {T1, T2}
+    if ct[1].shift == 0 && ct[2].scale == 1
+        print(io, "as𝕀")
+    else
+        print(io, "as(Real, ", ct[1].shift, ", ", ct[1].shift + ct[2].scale, ")")
+    end
+end
```
Also here, it's not the same as asI, so it shouldn't be printed in the same way.

Ickaser replied:
Fair point

devmotion commented on src/scalar.jl:
```diff
@@ -218,6 +356,52 @@ const asℝ = as(Real, -∞, ∞)
+        print(io, "as(Real, ", ct[1].shift, ", ", ct[1].shift + ct[2].scale, ")")
+    end
+end
+Base.show(io::IO, ct::CompositeScalarTransform{Tuple{TVLogistic}}) = print(io, "as𝕀")
```
AFAICT this should be removed when the definition of asI is fixed above:
```suggestion
```

devmotion commented on src/scalar.jl:
```diff
@@ -218,6 +356,52 @@ const asℝ = as(Real, -∞, ∞)
+        print(io, "as(Real, ", ct[1].shift, ", ∞)")
+    end
+end
+Base.show(io::IO, ct::CompositeScalarTransform{Tuple{TVExp}}) = print(io, "asℝ₊")
```
Not necessary when the definition of `TVExp` is fixed above:
```suggestion
```

Ickaser commented:
I appreciate the suggestions @devmotion! Many of your suggestions are on things that result from my inexperience so I'm happy to be learning better and more careful technique. To address some of the overarching questions:
1. I went ahead and included `inverse_and_logjac` in this PR since it was already implemented for `ScaledShiftedLogistic`, which I am replacing with `TVScale`, `TVShift`, and `TVLogistic`, so the only really new part would be `inverse_and_logjac` for the `TVExp` and `TVNeg` transforms. I didn't see obvious reasons it needed to keep its WIP status and tried to include it in tests. That said, I am not committed to its functionality and don't mind moving it to another PR using this one as a basis.
2. The odd definitions of `as𝕀` and `asℝ₊` and the weird print methods are the result of my trying to keep the previous string printing behavior. If we have `asℝ₊ = TVExp()`, then the string output of e.g. `TVScale(5.0) ∘ TVExp()` is `TVScale(5.0) ∘ asℝ₊` which feels confusing to me, so I decided to wrap the single transform in the composition. 
  - Likewise for the `as(Real, 0, \infty)` getting printed as `asR+` even though it's not strictly equal. This part I'm less committed to.
3. Some of the `@argcheck`s are me trying to follow the previous style, I don't mind dropping those.

Other stuff I'll respond in the individual threads

Ickaser commented:
With respect to the `inverse` methods all accepting `Number` My motivation for getting involved and building this PR is that I want to have transforms that add units by scaling, e.g. `TVScale(5u"m")`. So at the very least, the inverse of `TVScale` needs to accept a broad enough type to allow Unitful quantities, which are `<:Number` but not `<:Real`; in principle this can mean that on the inverse transform we get numbers that are not pure floats (although in the case of Unitful, when the units exactly cancel we get a plain float back).

Ickaser commented on src/scalar.jl:
```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
+
+Base.getindex(t::CompositeScalarTransform, i) = t.transforms[i]
+Base.firstindex(t::CompositeScalarTransform) = firstindex(t.transforms)
+Base.lastindex(t::CompositeScalarTransform) = lastindex(t.transforms)
```
That's a good point, this is the kind of thing that feels cool and exciting to me (index into my custom type! awesome) but even the `firstindex` and `lastindex` were examples of the interface feeling incomplete

devmotion commented:
> If we have asℝ₊ = TVExp(), then the string output of e.g. TVScale(5.0) ∘ TVExp() is TVScale(5.0) ∘ asℝ₊

That's the expected behaviour IMO, it should print `TVScale(5.0) ∘ asℝ₊`. It happens everywhere in the Julia ecosystem, e.g. `Vector{T}` is not a separate type but just an alias for `Array{T,1}`.

tpapp commented on src/scalar.jl:
```diff
@@ -52,7 +53,144 @@ transform(::Identity, x::Real) = x
+transform(t::TVScale, x::Real) = t.scale * x
+transform_and_logjac(t::TVScale{T}, x::Real) where {T<:Real} = transform(t, x), log(t.scale) 
+
+inverse(t::TVScale, x::Number) = x / t.scale
```
My understanding is that this is motivated by supporting `Unitful` and similar. It is fine with me.

tpapp commented:
@devmotion, thanks for the review. As for the following:

> What impact does this change have on compilation time?

I did not benchmark, but I am OK with a minor hit in compilation time in exchange for more flexibility. Eg I want to introduce other R->R+ transforms, as exp is not always the best option.

> What impact does this change have on performance?

Benchmarks above suggest it is innocuous.

> What impact does this change have on numerical accuracy?

It just does what those transformations did previously, only in composable steps.

Ickaser commented:
- With the change to make `asR+ = TVExp()` and `asI = TVLogistic()`, this leaves `asR- isa CompositeScalarTransform{Tuple{TVNeg, TVExp}} `. So that won't get printed as a single element transform. This seems fine to me.
- `TransformVariables.compose` is now just an alias to `CompositionsBase.compose`; I have not exported `compose` but we could mark it with `public`.
- The old implementation of `ShiftedExp` and `ScaledShiftedLogistic` is gone, including from the tests and docs
- I added some docs explaining how one would add new methods for `TVScale` that create number types where multiplying/dividing isn't desired
- We could conceivably add some sort of convenience constructor like `as(Real, 5u"m", \infty)` that would automatically add units. But at the moment I can't think of a non-awkward syntax for that so I'll let that go for now, maybe make a future PR if I hit on something good.

Ickaser commented:
After trying to use this in my code base, the two things I found myself wanting were:
- `as(Real, 1u"m", 5u"m")`
- A `ScalarTransform` that allows for a variable to be wrapped in some other function (or output type), e.g.
```
struct TVConstWrap <: ScalarTransform end
TransformVariables.transform(::TVConstWrap, x) = MyConstType(x)
TransformVariables.inverse(::TVConstWrap, x) = x.value
```
(although I did already add docs indicating this could be done by extending the `TVScale` transformation...)

That said, those can go into a future PR, and I'm content with this PR.

tpapp commented on docs/src/index.md:
```diff
@@ -98,6 +98,39 @@ asℝ₋
+This composition works with any scalar transform in any order, so `TVScale(4) ∘ as(Real, 2, ∞) ∘ TVShift(1e3)` is a valid transform.
+This is useful especially for making sure that values near 0, when transformed, yield usefully-scaled values for a given variable.
+
+In addition, the `TVScale` transform accepts arbitrary types. It can be used as the outermost transform (so leftmost in the composition) to add Unitful units to a number (or to create other exotic number types which can be constructed by multiplying, such as a `ForwardDiff.Dual`).
```
```suggestion
In addition, the `TVScale` transform accepts arbitrary types. It can be used as the outermost transform (so leftmost in the composition) to add, for example, `Unitful` units to a number (or to create other exotic number types which can be constructed by multiplying, such as a `ForwardDiff.Dual`).

However, note that calculating log Jacobian determinants may error for types that are not real numbers.
```

devmotion commented on src/scalar.jl:
```diff
@@ -50,90 +51,131 @@ struct Identity <: ScalarTransform end
+Base.:∘(t::ScalarTransform, ct::CompositeScalarTransform) = CompositeScalarTransform((t, ct.transforms...))
+Base.:∘(ct::CompositeScalarTransform, t::ScalarTransform) = CompositeScalarTransform((ct.transforms..., t))
+Base.:∘(ct1::CompositeScalarTransform, ct2::CompositeScalarTransform) = CompositeScalarTransform((ct1.transforms..., ct2.transforms...))
+Base.:∘(t::ScalarTransform, tt::Vararg{ScalarTransform}) = CompositeScalarTransform((t, tt...))
```
Is this actually needed? It also seems wrong if a few of `tt` are `CompositeScalarTransform`. Based on how this expression is parsed, I assumed it's sufficient to define `∘` for two arguments:

```julia
julia> Meta.dump(:(f ∘ g ∘ h))
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: Symbol ∘
    2: Expr
      head: Symbol call
      args: Array{Any}((3,))
        1: Symbol ∘
        2: Symbol f
        3: Symbol g
    3: Symbol h
```

devmotion replied:
```suggestion
```

Ickaser replied:
I like the idea of allowing `compose(t1, t2, t3, t4)`, rather than forcing `compose(compose(compose(t1, t2), t3), t4)`, so that was my intention for this method. But you are right that it can give somewhat wrong results if there are `CompositeScalarTransform`s inside the `Vararg`--fixed this now with a `foldl` that will recursively call the two-argument forms, and added tests that catch this case.

devmotion commented on src/scalar.jl:
```diff
@@ -218,24 +261,31 @@ const asℝ = as(Real, -∞, ∞)
+# Single scalar transforms
+Base.show(io::IO, ::Identity) = print(io, "asℝ")
+Base.show(io::IO, ::TVExp) = print(io, "asℝ₊")
+Base.show(io::IO, ::TVLogistic) = print(io, "as𝕀")
```
```suggestion
```

Ickaser replied:
Same here as for `asR+`.

devmotion commented on test/runtests.jl:
```diff
@@ -6,9 +6,10 @@ using LogDensityProblemsAD
     AbstractTransform, ScalarTransform, VectorTransform, ArrayTransformation,
     unit_triangular_dimension, logistic, logistic_logjac, logit, inverse_and_logjac,
-    NOLOGJAC, transform_with
+    NOLOGJAC, transform_with 
```
```suggestion
    NOLOGJAC, transform_with
```

Ickaser commented on src/scalar.jl:
```diff
@@ -218,24 +261,31 @@ const asℝ = as(Real, -∞, ∞)
-    end
+# Single scalar transforms
+Base.show(io::IO, ::Identity) = print(io, "asℝ")
+Base.show(io::IO, ::TVExp) = print(io, "asℝ₊")
```
No, it's the other way around--without this show method both `asR+` and `TVExp()` print as `TVExp()`. Personally I prefer it showing as `TVExp()`, since I think it is more explicit, but in order to have any object that prints `asR+` (as in prior behavior) we have to include this, since `asR+ === TVExp()`.

Ickaser commented on src/scalar.jl:
```diff
@@ -218,24 +261,31 @@ const asℝ = as(Real, -∞, ∞)
-        print(io, "as(Real, -∞, ", t.shift, ")")
-    end
+# Single scalar transforms
+Base.show(io::IO, ::Identity) = print(io, "asℝ")
```
This show method already existed as is, I just moved it within the file to be next to the others.

tpapp commented:
@Ickaser, thanks for the extension and your patience in addressing all the comments!
